triplet/main.py

```diff
@@ -0,0 +1,184 @@
+    different class.
+    """
+    def __init__(self, *arg, **kw):
+        super(TripletMNIST, self).__init__(*arg, **kw)
+
+        print('Generating triplets ...')
+        self.n_triplets = args.n_triplets
+        self.train_triplets = self.generate_triplets(self.train_labels)
+
+    def generate_triplets(self, labels):
+        triplets = []
+        labels = labels.numpy()
+
+        ulabels = np.unique(labels)
+        matches, no_matches = dict(), dict()
+        for x in ulabels:
+            matches[x] = np.where(labels == x)[0]
+            no_matches[x] = np.where(labels != x)[0]
+
+        candidates = np.random.randint(0, len(labels), size=self.n_triplets)
+        candidates = labels[candidates]
+        for x in candidates:
+            idx_a, idx_p = np.random.choice(matches[x], 2, replace=False)
+            idx_n = np.random.choice(no_matches[x], 1)[0]
+            triplets.append([idx_a, idx_p, idx_n])
+        return np.array(triplets)
+
+    def __getitem__(self, index):
+        if self.train:
+            t = self.train_triplets[index]
+            a, p, n = self.train_data[t[0]], self.train_data[t[1]],\
+                      self.train_data[t[2]]
+
+        # doing this so that it is consistent with all other datasets
+        # to return a PIL Image
+        img_a = Image.fromarray(a.numpy(), mode='L')
+        img_p = Image.fromarray(p.numpy(), mode='L')
+        img_n = Image.fromarray(n.numpy(), mode='L')
+
+        if self.transform is not None:
+            img_a = self.transform(img_a)
+            img_p = self.transform(img_p)
+            img_n = self.transform(img_n)
+
+        return img_a, img_p, img_n
+
+    def __len__(self):
+        if self.train:
+            return self.train_triplets.shape[0]
+
+
+class TNet(nn.Module):
+    """TFeat model definition
+    """
+    def __init__(self):
+        super(TNet, self).__init__()
+        self.conv1 = nn.Conv2d(1, 32, kernel_size=7)
+        self.conv2 = nn.Conv2d(32, 64, kernel_size=5)
+        self.fc1 = nn.Linear(64*8*8, 128)
+
+    def forward(self, x):
+        x = F.tanh(self.conv1(x))
+        x = F.max_pool2d(x, 2)
+        x = F.tanh(self.conv2(x))
+        x = x.view(-1, 64*8*8)
+        x = F.tanh(self.fc1(x))
+        return x
+
+
+class TripletMarginLoss(nn.Module):
+    """Triplet loss function.
+    Based on: http://docs.chainer.org/en/stable/_modules/chainer/functions/loss/triplet.html
+    """
+    def __init__(self, margin):
+        super(TripletMarginLoss, self).__init__()
+        self.margin = margin
+
+    def forward(self, anchor, positive, negative):
+        dist = torch.sum(
+            (anchor - positive) ** 2 - (anchor - negative) ** 2,
+            dim=1) + self.margin
+        dist_hinge = torch.clamp(dist, min=0.0)  # maximum between 'dist' and 0.0
+        loss = torch.mean(dist_hinge)
+        return loss
+
+
+def triplet_loss(input1, input2, input3, margin=1.0):
```

Comment:
Please paste that code into this function. Moving it out to these separate classes doesn't add anything.